### ISSUE TITLE
Consistent parallel AffineConstraints

### ISSUE THREAD

**Issue by tjhei** [dealii/dealii#6255]
Prompted by a problem on the mailing list I got reminded of a recurring problem:
We currently have no way to validate if the entries in a ConstraintMatrix are consistent between processors. This can generate very hard to find bugs.

My thought would be that the entries need to be identical for all stored lines (locally_relevant DoFs).

So I wrote a function ``ConstraintMatrix::is_consistent_in_parallel`` (see #6254). It works and detects for example if you mess with boundary ids only locally. What I did not expect, that a simple adaptively refined grid also fails: some processors are missing hanging node constraints in their relevant set. (grid 3 in the test)

What now?
a) My logic is incorrect and I need to relax the check in the new function.
b) make_hanging_node_constraints is broken
c) we need to always communicate constraints (how exactly?)

**Comment by tjhei** [comment]
details about the problem:
```
Proc 1 got line 110 from 0 wrong values!
Proc 1 got line 111 from 0 wrong values!
Proc 1 got line 114 from 0 inhomogeneity 0 != -5.25
Proc 1 got line 115 from 0 inhomogeneity 0 != -5.25
```
This means that proc 0 has 110,111,114,115 in her relevant set but doesn't know about the constraints. Are we just lucky that things work?

**Comment by bangerth** [comment]
How does grid 3 look like? And what constraints are missing?

It's conceivable that we could build more constraints though I'd be surprised if that was *necessary* -- just because we don't have all doesn't mean that we would get a different matrix. After all, when we assemble a linear system, we only ever touch locally *active* DoFs.

**Comment by kronbichler** [comment]
I would have said that the hanging node constraints in the ghost layer are not necessarily correct, namely on hanging nodes that are on a face that does not connect to the locally owned range (i.e., it is on the opposite end). So a relaxed checked would only look into the *locally active* DoFs for the equivalence.

So I think the hanging node constraints are computed correctly for what matters in typical computations.

**Comment by tjhei** [comment]
> does grid 3 look like? And what constraints are missing?

2d subdivided_hyper_cube with one coarsened and one refined cell (haven't looked at it). I assume it involves locally relevant and not active DoFs like @kronbichler suggests.

> So a relaxed checked would only look into the locally active DoFs for the equivalence.

Okay. Maybe I should look at that first. It still scares me though, because the missing entries could chain to locally active DoFs.

**Comment by bangerth** [comment]
On 04/17/2018 10:39 AM, Timo Heister wrote:
> Okay. Maybe I should look at that first. It still scares me though, 
> because the missing entries could chain to locally active DoFs.

But there is nothing you can do about it if the constraints are between 
a ghost and an artificial cell. The DoFs on the ghost side are locally 
relevant, but you can't compute their constraints.

**Comment by tjhei** [comment]
> But there is nothing you can do about it if the constraints are between
> a ghost and an artificial cell.

Well, you could communicate the constraints from a processor who knows about it (as we will need for hp in parallel).

**Comment by tjhei** [comment]
So, the good news: if I only check for consistency on locally active DoFs, my existing tests now pass. This means that hanging node constraints seem to be okay as on locally active DoFs.

The bad news: The example from the mailing list (mistake: assign boundary ids manually but only for locally owned cells) is no longer detected as inconsistent because wrong values are only present for locally relevant DoFs.

Should we just go ahead with #6254 and ignore this special case? Any other idea how I could detect this?

**Comment by bangerth** [comment]
On 04/18/2018 11:08 AM, Timo Heister wrote:
> The bad news: The example from the mailing list (mistake: assign 
> boundary ids manually but only for locally owned cells) is no longer 
> detected as inconsistent because wrong values are only present for 
> locally relevant DoFs.

Why did that happen? Did VectorTools::interpolate_b_v() accidentally add 
an entry into the ConstraintMatrix that should not have been there, or 
did it forget to put one in there?

The first case you should always be able to check -- if you have a 
constraint on a non-locally owned DoF, then that constraint must also be 
there on the processor that owns the DoF. The second case (forgetting 
something) is more difficult.

**Comment by tjhei** [comment]
> Why did that happen?

One processor got an incorrect boundary id for a face of a ghost cell (because set_boundary_id was only called for owned cells). Because ``interpolate_b_v`` is also executed for ghost faces, the value of one of the DoFs of that face (that happened to be locally owned) was set to an incorrect value. Does that make sense?

So this is not a matter of having vs not having an entry, it is an incorrect inhomogeneous value.

**Comment by bangerth** [comment]
On 04/18/2018 04:51 PM, Timo Heister wrote:
> One processor got an incorrect boundary id for a face of a ghost cell 
> (because set_boundary_id was only called for owned cells). Because 
> |interpolate_b_v| is also executed for ghost faces, the value of one of 
> the DoFs of that face (that happened to be locally owned) was set to an 
> incorrect value. Does that make sense?

Yes. Ugh.


> So this is not a matter of having vs not having an entry, it is an 
> incorrect inhomogeneous value.

I guess it was a vertex DoF. So this one DoF is then locally active on 
both processors, and they will have differing constraints for it. You 
should be able to check that: because the DoF is only locally owned on 
one processor, the processor where it is locally active but not locally 
owned should check with the owning processor whether the two are the same.

**Comment by tjhei** [comment]
> I guess it was a vertex DoF.

Yes.

> So this one DoF is then locally active on
> both processors, and they will have differing constraints for it

No, it is locally owned on one and locally relevant on the other processor (it is basically one vertex away from the one on the processor boundary).

**Comment by bangerth** [comment]
On 04/19/2018 11:14 AM, Timo Heister wrote:
>     So this one DoF is then locally active on
>     both processors, and they will have differing constraints for it
> 
> No, it is locally owned on one and locally relevant on the other 
> processor (it is basically one vertex away from the one on the processor 
> boundary).

Wait, this doesn't make sense. A vertex DoF needs to be locally owned on 
one processor and locally active on another, at least if it's shared 
between two cells. Or is it a corner vertex of the whole domain?

**Comment by tjhei** [comment]
as discussed in person:
this is not something we can detect in the ConstraintMatrix. We need to directly transfer boundary/manifold ids and check that they are consistent...

**Comment by tjhei** [comment]
No. I think ideally we would:
1. create a function to exchange_boundary/manifold_ids
2. check that those ids are consistent in some other function call (not sure exactly where)

Future plan, maybe?

**Comment by marcfehling** [comment]
I ran into this issue again while trying to solve a Stokes problem on the `pipe_junction` geometry with hp-refinement. I thought at first that there is something wrong with my setup, but solving the problem works fine with the check disabled. I was not able to create a minimal example that reproduces the problem so that we could add it to the testsuite, so instead I summarize what I found.

---

So after one adaptive p-refinement step on the described problem with 2 MPI processes, the `is_consistent_in_parallel` check failed after calling `close`. This problem also occurs on the same geometry with only p-refinement and `FE_Q` elements. It reports:
```
Proc 0 got line 12852 from 1 with wrong values!
Proc 0 got line 12853 from 1 with wrong values!
Proc 0 got line 12854 from 1 with wrong values!
Proc 0 got line 12968 from 1 with wrong values!
Proc 0 got line 12969 from 1 with wrong values!
Proc 0 got line 12970 from 1 with wrong values!
6 inconsistent lines discovered!
```

Also after calling `close`, I wrote the content of the `AffineConstraints` object of each processor to the filesystem. The lines related to the reported dofs (when I interpreted https://github.com/dealii/dealii/issues/6255#issuecomment-382039369 correctly) are identical on both processes. That left me confused.
```
constraints_proc0
=================
...
    12852 13058:  0.236379
    12852 13059:  0.0493352
    12852 13066:  0.879946
    12852 13067:  -0.165661
...

constraints_proc1
=================
...
    12852 13058:  0.236379
    12852 13059:  0.0493352
    12852 13066:  0.879946
    12852 13067:  -0.165661
...
```
I also uploaded the whole files [constraints_proc0.txt](https://github.com/dealii/dealii/files/8036474/constraints_proc0.txt), [constraints_proc1.txt](https://github.com/dealii/dealii/files/8036477/constraints_proc1.txt).

So is there something wrong with `is_consistent_in_parallel` itself? Or did I do something terribly wrong?

---

I could write a small failing test with a serialized geometry of the above scenario, but its 30k DoFs makes it difficult do debug. I will try to find a smaller example. In the meantime, I'll mark this for the 10.0 release milestone so that this won't slip below our radar.

**Comment by tjhei** [comment]
> So is there something wrong with `is_consistent_in_parallel` itself? Or did I do something terribly wrong? Issue could also be related to #8995.

This is precisely what I was going to say. I discovered differences in constraints in the order of 1e-16. We then removed the call to this function from ASPECT. I think if your problem gets sufficiently complex (curved geometry, manifolds, etc.), you can end up with slightly different constraints.
We could "fix" ``is_consistent_in_parallel`` by introducing a tolerance or, likely the better approach, actually communicate the ConstraintMatrix. With the new ConsensusAlgorithm this should be fairly easy to do.

**Comment by bangerth** [comment]
I would love to know where these 1e-16 differences actually originate, but I imagine it to be quite difficult to figure this out.

It is true that we could communicate constraints quite easily. We know which processes need to know about constraints on our own locally-owned DoFs, i.e., who we need to send information to. `MPI::some_to_some()` could easily facilitate this. Given how much trouble we have had with constraints over the years, this might actually be worth the time.

**Comment by peterrum** [comment]
@marcfehling You could also call `AffineConstraints::make_consistent_in_parallel()`; after that you should be on the safe side.

**Comment by kronbichler** [comment]
Is there anything more we plan to do on this topic before the release?

**Comment by kronbichler** [comment]
Are there any plans for the current release? Otherwise, I suggest to postpone the issue.

**Comment by tamiko** [comment]
Another year another bump to a different release :smile:

**Comment by kronbichler** [comment]
Given #17175 and the fix presented there, what do we expect to still do here? Are there more things we want to do in addition, e.g. in terms of the roundoff or inconsistent setting of boundary ids and the like, as initially discussed above,

> mistake: assign boundary ids manually but only for locally owned cells

and what verification we would do for the present release? We keep bumping the milestone target, but it feels we are at least having a solution that is much better than before, so it might make sense to finalize this issue soon?

**Comment by kronbichler** [comment]
As there seem to be no new comments here, let me close this as completed. We can open new issues with more specific problems if they arise again.
